lib/enterprise-event-stream.js

```diff
@@ -0,0 +1,252 @@
+
+	if (this._options.endDate) {
+		params.created_before = this._options.endDate;
+	}
+
+	if (this._options.eventTypeFilter) {
+		params.event_type = this._options.eventTypeFilter;
+	}
+
+	if (this._options.chunkSize) {
+		params.limit = this._options.chunkSize;
+	}
+
+	this._client.events.get(params, function(err, result) {
+		if (err) {
+			self.emit('error', err);
+			// If there was a "permanent" error, we would call the callback with it here.
+			// But it's not clear which errors are truly permanent?
+			// If Box is down or returning errors for an extended period, we still want to resume when it recovers.
+			// So, continue polling at the regular frequency.
+			// Don't use a shorter retry interval (to avoid DDOSing Box).
+		} else if (result && result.entries) {
+			// console.log(`Fetched events from ${options.streamPosition || options.createdAfter} (limit ${options.limit}) => ${result.entries.length} events (next ${result.next_stream_position})`);
+
+			if (result.entries.length > 0) {
+				// Only update the stream position if there were events returned.
+				// The API currently returns next_stream_position = 0 if there are no events (may be a bug?).
+				// But we don't want to start over at the beginning in that case, so ignore it.
+				self._streamPosition = result.next_stream_position;
+
+				// Notify the reader of the new stream position.
+				// Stream readers can respond to the 'streamPosition' event to persist the stream position.
+				self.emit('streamPosition', self._streamPosition);
+
+				callback(null, result.entries);
+				return;
+			}
+		}
+
+		if (!self._options.pollingInterval) {
+		// If polling is disabled, end the stream.
+			callback();
+			return;
+		}
+
+		// There were no events returned (or an error occurred), so schedule another poll.
+		const delay = self._options.pollingInterval * 1000;
+
+		// Stream readers can use this to flush buffered events to a downstream system.
+		self.emit('wait', delay);
+
+		setTimeout(function() {
+			self.fetchEvents(callback);
+		}, delay);
+	});
+};
+
+/**
+ * Implementation of the stream-internal read function.	This is called
+ * by the stream whenever it needs more data, and will not be called again
+ * until data is pushed into the stream.
+ * @returns {void}
+ * @private
+ */
+EnterpriseEventStream.prototype._read = function() {
+	// Fetch the next chunk of events.
+	const self = this;
+
+	// This will poll forever until events are available.
+	this.fetchEvents(function(err, events) {
+		if (err || !events || events.length === 0) {
+			// Close the stream if there was a "permanent" failure or we reached the end of the events.
+			self.push(null);
+			return;
+		}
+
+		// Pause the stream to avoid race conditions while pushing in the new events.
+		// Without this, _read() would be called again from inside each push(),
+		// resulting in multiple parallel calls to fetchEvents().
+		// See https://github.com/nodejs/node/issues/3203
+		const wasPaused = self.isPaused();
```

Comment:
I remember you mentioned this before; is this a problem for the User Events stream as well?